src/metabase/upload.clj

```diff
@@ -753,7 +754,12 @@
                             (map :id)
                             seq)]
     ;; Ideally we would do all the filtering in the query, but this would not allow us to leverage mlv2.
-    (model-persistence/invalidate! {:card_id [:in model-ids]})))
+    (model-persistence/invalidate! {:card_id [:in model-ids]})
+    ;; Also purge the metadata, so that newly added columns are visible, and types are updated.
+    (doseq [id model-ids]
+      (let [query    (t2/select-one-fn :dataset_query [:model/Card :dataset_query] id)
+            metadata (card.metadata/infer-metadata query)]
+        (t2/update! :model/Card id {:result_metadata metadata})))))
```

Comment:
This is to drastic I think - the problem is that there is all sorts of config that might have been saved on top of the inferred metadata.

Going to reduce the scope to only updating column types and adding missing columns.

Reply:
Went with adding and removing only, updating looks more treacherous.

Reply:
Wouldn't say that `refresh-metadata` is easy to read though 😁 good enough, but not amazing 😁